### PR TITLE
 Remove padding from scales for hipBLASlt calls

### PR DESCRIPTION
**Pull request by alextmagro**
Removes padding for scale vectors that are used mainly for MXFP8.

### CODE REVIEW

File: tests/cpp/operator/test_cublaslt_gemm.cu [hunk 1]
```diff
@@ -345,8 +347,11 @@ void performTest(const TestParams& params) {
     if (!has_fp8) {
       GTEST_SKIP() << "MXFP8 scaling mode requires Float8 types";
     }
-    if (params.m % 32 != 0 || params.n % 32 != 0 || params.k % 32 != 0) {
-      GTEST_SKIP() << "MXFP8 requires M, N, K to be multiples of 32";
+    if (params.m % 16 || params.n % 16) {
+      GTEST_SKIP() << "MXFP8 requires M & N to be multiples of 16";
+    }
+    if (params.k % 128) {
```
Comment (ipanfilo) [inline tests/cpp/operator/test_cublaslt_gemm.cu:352]: Is it hipblasLt limitation?

File: transformer_engine/common/comm_gemm_overlap/comm_gemm_overlap.cpp [hunk 2]
```diff
@@ -189,6 +189,7 @@ TensorWrapper CommOverlapCore::get_tensor_chunk(const TensorWrapper &source, siz
   NVTE_DIM_CHECK(chunk_height > 0 && chunk_width > 0, "Attempted to get empty tensor chunk");
   NVTE_DIM_CHECK(chunk_height <= height && chunk_width <= width,
                  "Attempted to get out-of-bounds tensor chunk");
+#ifndef __HIP_PLATFORM_AMD__
```
Comment (ipanfilo) [inline transformer_engine/common/comm_gemm_overlap/comm_gemm_overlap.cpp]: I think this file is not currently compiled for ROCm - it is for UB

File: transformer_engine/common/gemm/rocm_gemm.cu [hunk 3]
```diff
@@ -197,9 +197,59 @@ struct GemmParam {
...
+    float *scale_dptr = (float*)(scale_tensor.dptr);
+
+    const size_t total_elements = unpadded_rows * unpadded_cols;
+    const size_t threads        = 256;
+    const size_t blocks         = (total_elements + threads - 1) / threads;
+
+    unpad_mxfp8_scales_kernel<<<blocks, threads, 0, stream>>>
+                (scale_dptr, unpadded_rows, unpadded_cols, padded_cols);
+
+    NVTE_CHECK_CUDA(hipStreamSynchronize(stream));
```
Comment (ipanfilo) [inline transformer_engine/common/gemm/rocm_gemm.cu]: If we're using the same stream for GEMM, synchronize creates unnecessary bubble

File: transformer_engine/common/gemm/rocm_gemm.cu [hunk 4]
```diff
@@ -235,6 +285,9 @@ GemmParam CanonicalizeGemmInput(const transformer_engine::Tensor &A, const cubla
     // MXFP8
     // Note: Row-wise and column-wise data are scaled along different
     // dimensions (with matrix interpreted in row-major order).
+
+    unpad_mxfp8_checkpoint(A, is_A_transposed, m, n, k, stream);
+
```
Comment (ipanfilo) [inline transformer_engine/common/gemm/rocm_gemm.cu]: It is probably not the best place for this call - it does not update input parameters.
Also, what will be the code behaviour if we call GEMM twice with the same input - the first call will updad scale tensor and the second will do the same, right?

File: tests/cpp/operator/test_cublaslt_gemm.cu (hunk 1 above)
Comment (alextmagro) [reply tests/cpp/operator/test_cublaslt_gemm.cu:352]: Yes, these are the values that hipblastlt team provided to us. I tested just in case, but nothing smaller that 128 works for k.

File: transformer_engine/common/comm_gemm_overlap/comm_gemm_overlap.cpp (hunk 2 above)
Comment (alextmagro) [reply transformer_engine/common/comm_gemm_overlap/comm_gemm_overlap.cpp]: Yes, I can move it to the UB PR if you prefer?

File: transformer_engine/common/gemm/rocm_gemm.cu (hunk 4 above)
Comment (alextmagro) [reply transformer_engine/common/gemm/rocm_gemm.cu]: We talked about this during standup and the issues it would create in Python, I am currently working on moving the logic over there.

Comment (alextmagro) [reply transformer_engine/common/gemm/rocm_gemm.cu]: I have moved the check and unpadding to generic_gemm in pytorch. This provides a more robust way to check before calling the gemms, and directly modifies the pytorch tensors.

File: tests/cpp/operator/test_cublaslt_gemm.cu (hunk 1 above)
Comment (ipanfilo) [reply tests/cpp/operator/test_cublaslt_gemm.cu:352]: Is 32x128x32 config needed with 16x128x16 then?

File: tests/cpp/test_common.cu [hunk 5]
```diff
@@ -147,7 +147,11 @@ std::pair<scale_inv_meta, scale_inv_meta> get_scales(const NVTEShape& shape,
 
     scale_inv_meta ret_rowwise, ret_colwise;
 
+#ifndef __HIP_PLATFORM_AMD__
```
Comment (ipanfilo) [inline tests/cpp/test_common.cu]: Nit: swap to #ifdef to match other changes in the file

File: transformer_engine/common/comm_gemm_overlap/comm_gemm_overlap.cpp (hunk 2 above)
Comment (ipanfilo) [reply transformer_engine/common/comm_gemm_overlap/comm_gemm_overlap.cpp]: Yes, better move to UB because this file wold require more changes than those ifdefs, right?

File: tests/pytorch/test_sanity.py [hunk 6]
```diff
@@ -913,6 +913,78 @@ def test_sanity_fp8_gemm_with_unalignment(N, datatype):
     )
     torch.cuda.synchronize()
 
+@pytest.mark.skipif(not mxfp8_available, reason=reason_for_no_mxfp8)
+@pytest.mark.parametrize("N", [32])
+@pytest.mark.parametrize("K", [128])
+@pytest.mark.parametrize("M", [32])
```
Comment (ipanfilo) [inline tests/pytorch/test_sanity.py]: Better use non multiple of 32 to test this path is unpadding

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 7]
```diff
@@ -30,6 +36,42 @@ def validate_gemm_scale(scale: Optional[float], required: bool) -> float:
     return 0.0
 
 
+def unpad_scales(tensor: torch.Tensor, transpose: bool) -> torch.Tensor:
```
Comment (ipanfilo) [inline transformer_engine/pytorch/cpp_extensions/gemm.py]: I wonder if it can be called once when tensors are created?

Comment (alextmagro) [reply transformer_engine/pytorch/cpp_extensions/gemm.py]: When a user creates a tensor for the first time, we don't do padding to begin with -- This logic is for loading NV checkpoints only. I was thinking that when we load a pytorch checkpoint, the tensors are filled without calling the initializers, potentially missing the logic. Is there a way to guarantee the unpadding function is called when a Tensor is loaded?

File: tests/pytorch/test_sanity.py (hunk 6 above)
Comment (alextmagro) [reply tests/pytorch/test_sanity.py]: We require block sizes of 32 at the python level, so not possible to do a non-multiple. We are padding scales, so we will see a rowwise scale of (1,4) padded to (128,4), and a colwise scale of (4,1) being padded to (4,128).

File: transformer_engine/common/comm_gemm_overlap/comm_gemm_overlap.cpp (hunk 2 above)
Comment (alextmagro) [reply transformer_engine/common/comm_gemm_overlap/comm_gemm_overlap.cpp]: Yes, makes sense.

File: tests/cpp/operator/test_cublaslt_gemm.cu (hunk 1 above)
Comment (alextmagro) [reply tests/cpp/operator/test_cublaslt_gemm.cu:352]: I would say it makes sense to keep. This allows us to test a TE acceptable size with 32 while also ensuring unpadding and hipBLASlt is working with 16.

Comment (ipanfilo) [reply tests/cpp/operator/test_cublaslt_gemm.cu:352]: In this case I'd change 32x128x32 to 32x128x16 to test they work together

File: transformer_engine/pytorch/cpp_extensions/gemm.py (hunk 7 above)
Comment (ipanfilo) [reply transformer_engine/pytorch/cpp_extensions/gemm.py]: I think __init__ should be called when checkpointing is loading. Or loading can be intercepted be overriding load_from_state_dict

**Review by ipanfilo — APPROVED** [review]

File: tests/pytorch/references/blockwise_quantizer_reference.py [hunk 8]
```diff
@@ -1,3 +1,5 @@
+# This file was modified for portability to AMDGPU
+# Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (wangye805) [inline tests/pytorch/references/blockwise_quantizer_reference.py:2]: This is for blockwise fp8 recipe which we does not support yet

File: transformer_engine/pytorch/tensor/float8_blockwise_tensor.py [hunk 9]
```diff
@@ -1,3 +1,5 @@
+# This file was modified for portability to AMDGPU
+# Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (wangye805) [inline transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:2]: I don't think we support this blockwise fp8 yet
